activerecord/lib/active_record/query_logs.rb

```diff
@@ -0,0 +1,171 @@
+# frozen_string_literal: true
+
+require "active_support/core_ext/module/attribute_accessors_per_thread"
+
+module ActiveRecord
+  # Maintains a user-defined context for all queries and constructs an SQL comment by
+  # calling methods listed in +tags+
+  #
+  # Additional information can be added to the context to be referenced by methods
+  # defined in framework or application initializers.
+  #
+  # To add new comment tags, add a hash to the tags array containing the keys and values you
+  # want to add to the comment. Dynamic content can be created by setting a proc or lambda value in a hash,
+  # and can reference any value stored in the +context+ object.
+  #
+  # Example:
+  #
+  #    tags = [
+  #      :application,
+  #      { custom_tag: -> { context[:controller].controller_name } }
+  #    ]
+  #    ActiveRecord::QueryLogs.tags = tags
+  #
+  # Default tags available for use:
+  #
+  # * +application+
+  # * +pid+
+  # * +socket+
+  # * +db_host+
+  # * +database+
+  # * +line+ (reported via BacktraceCleaner)
+  #
+  # _When included in Rails, ActiveController and ActiveJob tags are also defined._
+  #
+  # * +controller+
+  # * +action+
+  # * +job+
+  #
+  # If required due to log truncation, comments can be prepended to the query instead:
+  #
+  #    ActiveRecord::QueryLogs.prepend_comment = true
+  #
+  # For applications where the query log tag content will not change during the lifetime of
+  # the request or job execution, the query log tag comment can be cached for reuse in every tagged query:
+  #
+  #    ActiveRecord::QueryLogs.cache_query_log_tags = true
+  #
+  # This option can be set during application configuration or in a Rails initializer:
+  #
+  #    config.active_record.cache_query_log_tags = true
+  #
+  # Clearing the cached SQL comment is possible by calling:
+  #
+  #    ActiveRecord::QueryLogs.clear_comment_cache!
+  module QueryLogs
+    mattr_accessor :taggings, instance_accessor: false, default: {}
+    mattr_accessor :tags, instance_accessor: false, default: [ :application ]
+    mattr_accessor :prepend_comment, instance_accessor: false, default: false
+    mattr_accessor :cache_query_log_tags, instance_accessor: false, default: false
+    mattr_accessor :backtrace_cleaner, instance_accessor: false, default: ActiveSupport::BacktraceCleaner.new
+    thread_mattr_accessor :cached_comment, instance_accessor: false
+
+    class << self
+      @@backtrace_cleaner.add_silencer { |line| line.match?(/lib\/active_(record|support)/) }
+
+      # Updates the context used to construct the query log tags. If a block is given,
+      # the context is updated temporarily.
+      # Resets the cached comment if <tt>cache_query_log_tags</tt> is +true+.
+      def set_context(**options)
+        original_context = context.dup
+        context.merge!(**options.symbolize_keys)
+        self.cached_comment = nil
+        if block_given?
+          yield
+          context.merge! original_context
+        end
+      end
```

Comment:
We need to an `ensure` here such that actions we wrap won't leak values in the `context`. I was thinking `set_context` would be an only block version and then `update` would be the non-block version.

```suggestion
      def set_context(**context)
        update(context)
        yield
      ensure
        update(context.transform_values! { nil })
      end
```

Reply:
Oh wow that's a really nice way of handling the reset - TIL `transform_values!` - excellent :)

Reply:
I've split these up now, but I thought that the previous version would have solved context leaking via

```
          yield
          context.merge! original_context
```

I'm happy with either approach though :)

Reply:
> I've split these up now, but I thought that the previous version would have solved context leaking via

It would've needed an `ensure` to guard against leaking in case of an exception being raised during the `yield`, but yes other than that 😊